subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.enterprise;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.function.Supplier;
+
+/**
+ * Runs submitted jobs in the background. The implementation is provided by Gradle.
+ */
+public interface GradleEnterprisePluginBackgroundJobExecutor extends Executor {
+    /**
+     * Execution traits specify additional flavors for the job execution.
+     * For example, a specific trait can signal that configuration cache inputs should not be recorded for the job.
+     */
+    interface ExecutionTraits {}
```

Comment:
From my naive and context-lacking perspective, this construct looks overkill. 

Might it be better to have `GradleEnterprisePluginBackgroundJobExecutor` not is-a `Executor` but instead `has-a` two different `Executor`s that behave with the traits? That would cut down the API surface quite a bit.

Reply:
I think it would be better to avoid the implementation details here and expose more “logical” executors with the right behaviour. 

Along those lines, could that be handled by the Gradle runtime? Gradle knows what the thread is doing when background work is submitted. It could use that context to do the appropriate decoration.

Reply:
> I think it would be better to avoid the implementation details here and expose more “logical” executors with the right behaviour.

Thanks, I see your point. This also opens some leeway in changing the properties of the underlying executor. I guess I shouldn't attempt to overgeneralize this API, especially considering it is an interim solution.

> Along those lines, could that be handled by the Gradle runtime? Gradle knows what the thread is doing when background work is submitted. It could use that context to do the appropriate decoration.

I'm not sure if I fully understand your idea here. Is it about deducing the traits automatically from the context? I think this couples the plugin and the core a bit too much.